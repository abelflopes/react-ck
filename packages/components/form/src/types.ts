import type { InputProps } from "@react-ck/input";
import type { SelectProps } from "@react-ck/select";
import type { TextareaProps } from "@react-ck/textarea";

type FieldTypes = "input" | "select" | "textarea";

// Utility type to compose fields, applies a type discriminator
// to easily distinguish fields
interface Field<T extends FieldTypes, P> {
  type: T;
  props: P;
}

// Types of fields available for the developer
type BaseFieldListMap<K extends Record<FieldTypes, unknown>> = K;
type FieldListMap = BaseFieldListMap<{
  input: Field<"input", InputProps>;
  select: Field<"select", SelectProps>;
  textarea: Field<"textarea", TextareaProps>;
}>;

// Utility type grouping all fields
type AnyField = FieldListMap[FieldTypes];

// Utility type to properly define value types according to its field
type BaseTypeValueMap<K extends Record<FieldTypes, unknown>> = K;
type TypeValueMap = BaseTypeValueMap<{
  input: NonNullable<InputProps["value"]> | null;
  select: NonNullable<SelectProps["value"]> | null;
  textarea: NonNullable<TextareaProps["value"]> | null;
}>;

// Utility type describing map provided by the developer
export type FormFieldMap = Record<string, AnyField>;

// Utility type describing the form payload
export type FormValues<T extends FormFieldMap> = { [key in keyof T]: TypeValueMap[T[key]["type"]] };

// Type describing form validator functions mapped to each field
export type FormValidators<T extends FormFieldMap> = {
  [key in keyof T]?: (values: FormValues<T>) => string | undefined;
};

export interface FormValidity<T extends FormFieldMap> {
  valid: boolean;
  fields: {
    [key in keyof T]?: {
      valid: boolean;
      error: string | undefined;
    };
  };
}
