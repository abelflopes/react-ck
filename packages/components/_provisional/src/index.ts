export * from "./image";

export * from "./avatar";

export * from "./dropdown";

export * from "./dock";

export * from "./breadcrumbs";

export * from "./scrollable-container";

export * from "./position-engine";

export * from "./top-bar";

export * from "./snackbar";

export * from "./overlay-spinner";

export * from "./menu";

export * from "./tooltip";

export * from "./popover";

export * from "./chat";

export * from "./file-uploader";

export * from "./attachment";

export * from "./snippet";
