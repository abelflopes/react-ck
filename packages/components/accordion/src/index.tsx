import React from "react";
import { List, type ListProps } from "@react-ck/list";
import { Collapse } from "@react-ck/collapse";
import styles from "./styles/index.module.scss";

export interface AccordionItem {
  header: NonNullable<React.ReactNode>;
  children: NonNullable<React.ReactNode>;
}

export interface AccordionProps extends Omit<ListProps, "items"> {
  items: AccordionItem[];
}

/**
 * An accordion is a vertically stacked list of interactive items. Each item can be "collapsed" with only a summary visible or “expanded” to show the full content for that item.
 * @param props - {@link AccordionProps}
 * @returns a React element
 */

export const Accordion = ({
  items,
  ...otherProps
}: Readonly<AccordionProps>): React.ReactElement => (
  <List
    {...otherProps}
    items={items.map(({ header, children }) => (
      <Collapse key={String(JSON.stringify(header))} header={header}>
        <div className={styles.content}>{children}</div>
      </Collapse>
    ))}
  />
);
