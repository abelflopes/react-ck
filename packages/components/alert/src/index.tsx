import React from "react";
import classNames from "classnames";
import styles from "./styles/index.module.scss";
import { Text } from "@react-ck/text";

export interface AlertProps extends React.HTMLAttributes<HTMLDivElement> {
  heading?: string;
  /** Specifies the visual style of the alert  */
  skin?: "neutral" | "primary" | "negative" | "average" | "positive";
}

// TODO: add dismiss

/**
 * Alert is a short message that provides contextual feedback in a prominent way.
 * @param props - {@link AlertProps}
 * @returns a React element
 */

export const Alert = ({
  heading,
  skin = "neutral",
  children,
  className,
  ...otherProps
}: Readonly<AlertProps>): React.ReactElement => (
  <div {...otherProps} className={classNames(styles.root, styles[skin], className)}>
    {heading && (
      <Text type="h4" as="p" variation="bold" margin={false} className={styles.heading}>
        {heading}
      </Text>
    )}
    {children}
  </div>
);
