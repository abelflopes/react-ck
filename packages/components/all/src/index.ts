export * from "./text";

export * from "./divider";

export * from "./banner";

export * from "./card";

export * from "./container";

export * from "./chip";

export * from "./collapse";

export * from "./list";

export * from "./overlay";
