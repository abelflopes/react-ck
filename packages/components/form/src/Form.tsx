import React, { useEffect, useState } from "react";
import {
  type InternalValues,
  type FormFieldMap,
  type FormValidators,
  type FormValidity,
  type FormValues,
} from "./types";
import classNames from "classnames";
import * as styles from "./styles/index.module.scss";
import { InputAdapter } from "./adapters/Input";
import { TextareaAdapter } from "./adapters/Textarea";
import { SelectAdapter } from "./adapters/Select";

const RESTRICTED_PROPS = ["selectedValue", "defaultValue", "value"];

export interface FormProps<T extends FormFieldMap>
  extends Omit<React.HTMLAttributes<HTMLDivElement>, "onChange"> {
  fields: T;
  onChange: (data: FormValues<T>) => void;
  onValidityChange?: (validity: FormValidity<T>) => void;
  values: FormValues<T>;
  validators?: FormValidators<T>;
  children?: React.ReactNode;
}

export const Form = <T extends FormFieldMap>({
  fields,
  onChange,
  onValidityChange,
  values,
  validators,
  className,
  children,
  ...props
}: Readonly<FormProps<T>>): React.ReactElement => {
  const [internalValues, setInternalValues] = useState<InternalValues<T>>({
    values,
    changedField: undefined,
    changeType: "internal",
  });
  const [fieldElements, setFieldElements] = useState<React.ReactNode[]>([]);
  const [validity, setValidity] = useState<FormValidity<T>>({
    valid: false,
    fields: {},
  });
  // State that informs that the values were changed by the user
  const [isTouched, setIsTouched] = useState<{ [key in keyof T]?: boolean }>({});

  // Create UI elements to render
  useEffect(() => {
    const tmpFieldElements: React.ReactNode[] = [];

    for (const fieldKey in fields) {
      if (fields[fieldKey] !== undefined) {
        const item = fields[fieldKey];

        if (item === undefined) continue;

        // Validate props
        if (Object.keys(item.props).some((i) => RESTRICTED_PROPS.includes(i))) {
          throw new Error(
            `Properties "${RESTRICTED_PROPS.join(
              ",",
            )}" are restricted, please use Form "defaultValues"`,
          );
        }

        const value = internalValues.values[fieldKey];

        const adapterProps = {
          key: fieldKey,
          fieldKey,
          value,
          isTouched: Boolean(isTouched[fieldKey]),
          valid: Boolean(validity.fields[fieldKey]?.valid),
          error: validity.fields[fieldKey]?.error,
          setInternalValues,
        };

        tmpFieldElements.push(
          ((): React.ReactNode => {
            switch (item.type) {
              case "input":
                return <InputAdapter {...{ ...item, ...adapterProps }} />;
              case "select":
                return <SelectAdapter {...{ ...item, ...adapterProps }} />;
              case "textarea":
                return <TextareaAdapter {...{ ...item, ...adapterProps }} />;
              default:
                throw new Error(`Field "${JSON.stringify(item)}" is not implemented`);
            }
          })(),
        );
      }
    }

    setFieldElements(tmpFieldElements);
  }, [fields, internalValues, isTouched, validity]);

  // Invoke onchange when internal state changes
  useEffect(() => {
    if (internalValues.changeType !== "user") return;
    onChange(internalValues.values);
  }, [internalValues, onChange]);

  // Update internal values when external state changes
  useEffect(() => {
    setInternalValues({
      values,
      changeType: "internal",
      changedField: undefined,
    });
  }, [values]);

  // Validate values
  useEffect(() => {
    if (!validators) return;

    setValidity({
      valid: true,
      fields: {},
    });

    for (const key in internalValues.values) {
      const error = validators[key]?.(internalValues.values);

      setValidity((v) => ({
        valid: !error && v.valid,
        fields: {
          ...v.fields,
          [key]: {
            valid: !error,
            error,
          },
        },
      }));
    }
  }, [internalValues.values, validators]);

  // Emit validity state change
  useEffect(() => {
    onValidityChange?.(validity);
  }, [onValidityChange, validity]);

  // Update state that informs that the values were changed by the user
  useEffect(() => {
    if (internalValues.changeType !== "user") return;

    setIsTouched((v) =>
      internalValues.changedField
        ? {
            ...v,
            [internalValues.changedField]: true,
          }
        : v,
    );
  }, [internalValues.changeType, internalValues.changedField]);

  return (
    <div {...props} className={classNames(className, styles.root)}>
      {fieldElements}
      {children}
    </div>
  );
};
