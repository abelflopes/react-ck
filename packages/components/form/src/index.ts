export * from "./Form";

export * from "./use-form";

export * from "./types";
