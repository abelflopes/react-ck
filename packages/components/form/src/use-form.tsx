import React, { useCallback, useEffect, useMemo, useState } from "react";
import { Form } from "./Form";
import type { FormValues, FormValidity, FormValidators, FormFieldMap } from "./types";

export interface UseFormProps<T extends FormFieldMap> {
  fields: T;
  initialValues: FormValues<T>;
  validators?: FormValidators<T>;
}

export interface UseFormData<T extends FormFieldMap> {
  form: React.ReactElement;
  initialValues: FormValues<T>;
  values: FormValues<T>;
  setValues: (values: FormValues<T>) => void;
  validity: FormValidity<T>;
}

export const useForm = <T extends FormFieldMap>({
  fields,
  initialValues,
  validators,
}: UseFormProps<T>): UseFormData<T> => {
  // Manually memorize in case initial values are passed directly in the hook to prevent infinite renders
  const [memorizedInitialValues, setMemorizedInitialValues] =
    useState<FormValues<T>>(initialValues);

  const [internalValues, setInternalValues] = useState<FormValues<T>>(memorizedInitialValues);
  const [validity, setValidity] = useState<FormValidity<T>>({
    valid: false,
    fields: {},
  });

  const form = useMemo<React.ReactElement>(
    () => (
      <Form
        fields={fields}
        values={internalValues}
        validators={validators}
        onValidityChange={setValidity}
        onChange={(v) => {
          if (JSON.stringify(v) === JSON.stringify(internalValues)) return;
          setInternalValues(v);
        }}
      />
    ),
    [fields, internalValues, validators],
  );

  const setValues = useCallback<UseFormData<T>["setValues"]>((v) => {
    setInternalValues(v);
  }, []);

  useEffect(() => {
    if (JSON.stringify(initialValues) === JSON.stringify(memorizedInitialValues)) return;
    setMemorizedInitialValues(initialValues);
  }, [initialValues, memorizedInitialValues]);

  return {
    form,
    initialValues: memorizedInitialValues,
    values: internalValues,
    setValues,
    validity,
  };
};
