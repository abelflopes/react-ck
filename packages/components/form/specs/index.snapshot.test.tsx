import React from "react";
import { Form } from "../src/index";
import renderer, { act, type ReactTestRenderer } from "react-test-renderer";
import * as LoginForm from "../fixtures/login";

describe("Snapshot Form", () => {
  test("renders correctly", async () => {
    let component: ReactTestRenderer | undefined;

    await act(() => {
      component = renderer.create(
        <Form
          fields={LoginForm.fields}
          validators={LoginForm.validators}
          values={LoginForm.values}
          onChange={() => undefined}
        />,
      );
    });

    expect(component?.toJSON()).toMatchSnapshot();
  });
});
