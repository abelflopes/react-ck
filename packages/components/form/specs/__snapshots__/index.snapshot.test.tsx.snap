// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Form renders correctly 1`] = `
<div
  className="root"
>
  <div
    className="root default"
  >
    <div
      className="content"
    >
      <input
        className="root"
        onChange={[Function]}
        placeholder="Email Address"
        value=""
      />
    </div>
  </div>
  <div
    className="root default"
  >
    <div
      className="content"
    >
      <input
        className="root"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
  </div>
</div>
`;
